File: 02. Algorithms/03. Searching/Problems/15. 278. First Bad Version/Solution.java
// https://leetcode.com/problems/first-bad-version/

/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int start = 0, end = n;
        while(start < end) {
            int mid = start + (end - start)/2;
            if(isBadVersion(mid) == false)
                start = mid + 1;
            else
                end = mid;
        }
        return start;
    }
}

File: 02. Algorithms/03. Searching/Problems/15. 278. First Bad Version/Solution.js
/**
 * Binary Search
 * Definition for isBadVersion()
 * https://leetcode.com/problems/first-bad-version/
 * 
 * Time O(log(n)) | Space O(1)
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 *     ...
 * };
 */

/**
 * @param {function} isBadVersion()
 * @return {function}
 */
var solution = function(isBadVersion) {
    /**
     * @param {integer} n Total versions
     * @return {integer} The first bad version
     */
    return function(n) {
        let left = 1;
        let right = n;

        let firstBad = 0;
        while(left <= right) {
            const mid = left + Math.floor((right-left)/2);
            const isBad = isBadVersion(mid);
            if(isBad) {
                firstBad = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return firstBad;
    };
};
